wdl-grammar/src/v1/lint/snake_case.rs

```diff
@@ -0,0 +1,191 @@
+//! Workflows, tasks and variables should be in lowered snake case.
+
+use std::collections::VecDeque;
+
+use convert_case::Case;
+use convert_case::Casing;
+use nonempty::NonEmpty;
+use pest::iterators::Pair;
+use wdl_core::concern::code;
+use wdl_core::concern::lint;
+use wdl_core::concern::lint::Group;
+use wdl_core::concern::lint::Rule;
+use wdl_core::concern::Code;
+use wdl_core::file::Location;
+use wdl_core::Version;
+
+use crate::v1;
+
+/// Detects names that should use snake case.
+///
+/// Workflows, tasks and variables should be declared
+/// using snake case.
+#[derive(Debug)]
+pub struct SnakeCase;
+
+impl<'a> SnakeCase {
+    /// Creates a warning for identifiers that are not proper snake case.
+    fn not_snake_case(&self, warning: SnakeCaseWarning<'_>) -> lint::Warning
+    where
+        Self: Rule<&'a Pair<'a, v1::Rule>>,
+    {
+        lint::warning::Builder::default()
+            .code(self.code())
+            .level(lint::Level::Low)
+            .group(self.group())
+            .subject("missing snake case")
+            .body("Identifier must be formatted using snake case.")
+            .push_location(warning.location)
+            .fix(format!(
+                "Replace {0} by {1}",
+                warning.identifier, warning.cased_identifier
+            ))
+            .try_build()
+            .unwrap()
+    }
+}
+
+/// Arguments for the `not_snake_case` function.
+struct SnakeCaseWarning<'a> {
+    /// Location of the warning.
+    location: Location,
+
+    /// Original identifier
+    identifier: &'a str,
+
+    /// Properly cased identifier
+    cased_identifier: &'a str,
+}
+
+impl Rule<&Pair<'_, v1::Rule>> for SnakeCase {
+    fn code(&self) -> Code {
+        Code::try_new(code::Kind::Warning, Version::V1, 6).unwrap()
+    }
+
+    fn group(&self) -> lint::Group {
+        Group::Naming
+    }
+
+    fn check(&self, tree: &Pair<'_, v1::Rule>) -> lint::Result {
+        let mut warnings = VecDeque::new();
+
+        for node in tree.clone().into_inner().flatten() {
+            if [
+                crate::v1::Rule::task_name,
+                crate::v1::Rule::workflow_name,
+                crate::v1::Rule::bound_declaration_name,
+                crate::v1::Rule::unbound_declaration_name,
+            ]
+            .contains(&node.as_rule())
+            {
+                let identifier: &str = node.as_span().as_str();
+                let cased_identifier: &str = &node.as_span().as_str().to_case(Case::Snake);
+                if identifier != cased_identifier {
+                    let warning = SnakeCaseWarning {
```

Comment:
I'd just embed the warning directly in the line 90 rather than make it a variable here (it doesn't appear to be used anywhere else).